### PR TITLE
BUG: DataFrame[dt64].where downcasting

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -262,7 +262,7 @@ def maybe_downcast_to_dtype(result: ArrayLike, dtype: str | np.dtype) -> ArrayLi
                 dtype = "int64"
             elif inferred_type == "datetime64":
                 dtype = "datetime64[ns]"
-            elif inferred_type == "timedelta64":
+            elif inferred_type in ["timedelta", "timedelta64"]:
```
Comment (mroeschke) [inline pandas/core/dtypes/cast.py:265]: Off topic, but should the above elif include `"datetime"`?

File: pandas/tests/dtypes/cast/test_downcast.py [hunk 2]
```diff
@@ -34,6 +37,13 @@
             "int64",
             np.array([decimal.Decimal(0.0)]),
         ),
+        (
+            # GH#45837
+            np.array([Timedelta(days=1), Timedelta(days=2)], dtype=object),
+            "infer",
+            np.array([1, 2], dtype="m8[D]").astype("m8[ns]"),
+        ),
+        # TODO: similar for dt64, dt64tz, Period, Interval?
```
Comment (mroeschke) [inline pandas/tests/dtypes/cast/test_downcast.py:46]: xfail these cases?

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py:265]: tried it and it broke things, so firmly in the "maybe" column

File: pandas/tests/dtypes/cast/test_downcast.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/dtypes/cast/test_downcast.py:46]: not yet sure what the correct behavior is.  will make xfails once confident

**Review by mroeschke — APPROVED** [review]

**Comment by jreback** [comment]
hmm this didn't seem to run all of the checks. can you rebase

**Comment by jreback** [comment]
can you rebase

**Comment by jbrockmendel** [comment]
rebased + greenish

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**